Shared component code

msafi opened the pull request:
This PR moves shared component code to a new function called `processComponentChanges`. Centralizing this code will come in handy when we add the logic to deal with issues that are caused by mask such as 

```
00\\1 (111) 111-1111
```

This PR also changes the way the React component works with regard to initial `value`. It now takes `this.props.value` and sets it as the initial `value` without trying to conform it. I don't see why the user would require that the plain initial value be conformed. But if that is the case, `react-text-mask` [exposes the `conformToMask`](https://github.com/msafi/text-mask/blob/43689563b373536821e163532b79e0bdd0a5c6be/react/src/reactTextMask.jsx#L105) function, which can be used by the user to pass an initial conformed value. Not conforming the initial value simplifies the requirements of the components and makes sense to me. What do you think about that @browniefed?

(I should probably document that `conformToMask` is exposed by `react-text-mask`)

browniefed commented:
I disagree with not formatting on the initial.
I personally in some cases strip out data and pass back in stripped out data (example phone numbers I strip dashes from).

So I'm expecting the component and library to take an arbitrary input and apply that to what ever value I'm giving it.
The main issue (at least with the React stuff) is that you are operating mostly as an uncontrolled input. The only reason the component is actually working as intended is because when the mask changes/guide we reprocess whatever the next value is and run it through the new mask. There by updating the internal state of the new application.

A point I may or may not argue is we would technically just conform the value it mask in the render function. This would cause problems with double calls to the `conformToMask` to send back the appropriate value in the `onChange`.


TLDR; Moral of the story the values passed into React don't necessarily come formatted, opening up the `conformToMask` works because internal state of the value is held in the component instance. `conformToMask` will then be called twice for every render. Initial value , and subsequent times on render even though the component internal state will call the `conformToMask`

msafi commented:
Ok, I'll change it back so that it conforms initial value.

I'll update this PR tomorrow.

browniefed commented:
One thing I was thinking about was something like.

```
<TextMask
{...conformToMask(value, mask, guide)}
/>
```
That way we could theoretically remove any internal state. The only thing we'd have to manage was the cursorPosition but that could likely be done.

msafi commented:
@browniefed 

I've updated the PR to make all components conform initial value. I've also added type coercion and type checking for `value`. Please take a look when you get a chance.

Regarding:

```
<TextMask
{...conformToMask(value, mask, guide)}
/>
```

I'm not in favor of doing it this way because it requires a little bit of extra-wiring from the user and it doesn't look nice.
